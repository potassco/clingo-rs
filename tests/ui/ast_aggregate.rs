use clingo::*;

fn main() {
    let loc = ast::Location::default();
    let sym = Symbol::create_id("test", true).unwrap();
    let term1: ast::Term = ast::symbolic_term(&loc, &sym).unwrap().into();
    let term2: ast::Term = ast::symbolic_term(&loc, &sym).unwrap().into();
    let term3: ast::Term = ast::symbolic_term(&loc, &sym).unwrap().into();
    let term4: ast::Term = ast::symbolic_term(&loc, &sym).unwrap().into();

    let atom1 = ast::symbolic_atom(term1).unwrap();
    let lit = ast::basic_literal_from_symbolic_atom(&loc, ast::Sign::NoSign, atom1).unwrap();

    let atom2 = ast::symbolic_atom(term2).unwrap();
    let lit2: ast::Literal = ast::basic_literal_from_symbolic_atom(&loc, ast::Sign::NoSign, atom2)
        .unwrap()
        .into();

    let condition = vec![lit2];
    let cond = ast::conditional_literal(&loc, lit, &condition).unwrap();
    let elements = vec![cond];

    let guard_l = ast::guard(ast::ComparisonOperator::GreaterThan, term3).unwrap();
    let guard_r = ast::guard(ast::ComparisonOperator::GreaterThan, term4).unwrap();
    let agg = ast::aggregate(&loc, Some(guard_l), &elements, Some(guard_r)).unwrap();

    drop(elements);
    let _end = agg;
}
