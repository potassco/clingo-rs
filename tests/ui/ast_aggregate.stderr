error[E0505]: cannot move out of `elements` because it is borrowed
  --> tests/ui/ast_aggregate.rs:27:10
   |
21 |     let elements = vec![cond];
   |         -------- binding `elements` declared here
...
25 |     let agg = ast::aggregate(&loc, Some(guard_l), &elements, Some(guard_r)).unwrap();
   |                                                   --------- borrow of `elements` occurs here
26 |
27 |     drop(elements);
   |          ^^^^^^^^ move out of `elements` occurs here
28 |     let _end = agg;
   |                --- borrow later used here
